### PR TITLE
ADMINUI-2311 Volumes UI

### PR DESCRIPTION
**Pull request by andyYatskov**

### CODE REVIEW

File: lib/auth.js [hunk 1]
```diff
@@ -18,7 +18,8 @@ var LOCKED_TIME = 4 * 60 * 60;
 var userLockTimes = {};
 var userLoginAttempts = {};
 
-var MANTA_ENALBED = require('../etc/config.json')['manta'] || false;
+var config = require('../etc/config.json');
+var MANTA_ENALBED = config.manta || false;
```
Comment (YZaicev) [inline lib/auth.js]: MANTA_ENABLED

File: lib/auth.js [hunk 2]
```diff
@@ -189,10 +190,12 @@ Auth.authenticate = function authenticate(req, res, next) {
                 return next(new restify.ConflictError('User does not have permission to use Operations Portal.'));
             }
 
+            var isVolapiEnabled = !!(config.datacenters[req.dc] && config.datacenters[req.dc].volapi);
```
Comment (YZaicev) [inline lib/auth.js]: isVolApiEnabled

File: www/js/components/pages/volumes/index.jsx [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+        app.vent.trigger('settitle', 'volumes');
+    },
+    showForm: function () {
+        this.setState({form: true});
+    },
+    hideForm: function () {
+        this.setState({form: false});
+    },
+    handleSave: function () {
+        console.log('handleSave');
```
Comment (YZaicev) [inline www/js/components/pages/volumes/index.jsx]: console.log 😱

File: www/js/components/pages/volumes/list.jsx [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    onSync: function () {
+        this.setState({volumes: this.state.volumes, loading: false});
+    },
+    showVolume: function (volume) {
+        adminui.vent.trigger('showcomponent', 'volume', {volume: volume});
+    },
+    navigateToOwnerDetailsPage: function (user) {
+        adminui.vent.trigger('showcomponent', 'user', {user: user});
+    },
+    deleteVolume: function (model, e) {
```
Comment (YZaicev) [inline www/js/components/pages/volumes/list.jsx]: e?

File: www/js/components/pages/volumes/list.jsx [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+    renderVolume: function (model) {
+        var volume = model.toJSON();
+        volume.state = volume.state || 'unknown';
+        return (
+            <tr key={volume.uuid}>
+                <td className="status">
+                    <span className={'state ' + volume.state}>{volume.state}</span>
+                </td>
+                <td className="name">
+                    <a onClick={this.showVolume.bind(this, model)}>{volume.name ? volume.name : volume.uuid}</a>
```
Comment (YZaicev) [inline www/js/components/pages/volumes/list.jsx]: volume.name ? volume.name : volume.uuid -> volume.name || volume.uuid

File: www/js/components/pages/volumes/networks.jsx [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+            <h3>Networks</h3>
+            <NetworksList
+                collection={this.state.networks}
+                showActions={false} />
+        </div>;
+    }
+});
+
+
+module.exports = VolumeNetworksList;
```
Comment (YZaicev) [inline www/js/components/pages/volumes/networks.jsx]: add line

File: www/js/components/pages/volumes/volume-form.jsx [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+            if (this.state.userNetworks[selectedUser]) {
+                return self.setState({
+                    error: null,
+                    loading: false,
+                    networks: self.state.userNetworks[selectedUser]
+                });
+            }
+            var req = this.networks.fetch({params: {
+                provisionable_by: selectedUser
+              //  fabric: true
```
Comment (YZaicev) [inline www/js/components/pages/volumes/volume-form.jsx]: remove

File: lib/auth.js (hunk 1 above)
Comment (andyYatskov) [inline lib/auth.js]: ))

**Comment by YZaicev** [comment]
:ok: 🔢

File: less/volumes.less [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+  a.delete {color: @state-danger-text; text-decoration: none; font-weight: bold; visibility: hidden; float: right;}
+  tr:hover .actions a, .actions:hover a {visibility: visible;}
+}
+
+#page-volume {
+  #page-vm;
+}
+.networks-list {
+  #page-networks > .networks-list;
+}
```
Comment (dan-luk) [inline less/volumes.less]: EOF

File: lib/adminui.js [hunk 9]
```diff
@@ -38,6 +38,13 @@ function resume(req, res, next) {
     next();
 }
 
+function checkVolApiConfig(req, res, next) {
+    if (req.sdc[req.dc].hasOwnProperty('volapi')) {
+        return next();
+    }
+    next (new restify.ServiceUnavailableError('Cannot connect to VOLAPI service. Please verify configuration.'));
```
Comment (dan-luk) [inline lib/adminui.js]: remove extra parentheses

File: lib/volumes.js [hunk 10]
```diff
@@ -0,0 +1,55 @@
+
```
Comment (dan-luk) [inline lib/volumes.js]: ➖

File: lib/volumes.js [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+        res.send(volume);
+        return next();
+    });
+}
+
+function create(req, res, next) {
+    req.log.info(req.body, 'VOLAPI Create Request');
+    req.sdc[req.dc].volapi.post('/volumes', req.body, function (err, _req, _res, volume) {
+        if (err) {
+            return next(err);
```
Comment (dan-luk) [inline lib/volumes.js]: no log here?

File: www/js/components/localnav.jsx [hunk 12]
```diff
@@ -20,7 +20,7 @@ var SecondaryNav = React.createClass({
         handleMenuSelect: React.PropTypes.func.isRequired,
         active: React.PropTypes.string
     },
-    _clickedMenuItem: function(e) {
+    _clickedMenuItem:function (e) {
```
Comment (dan-luk) [inline www/js/components/localnav.jsx]: why remove space

File: www/js/components/pages/volumes/index.jsx [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+var utils = require('../../../lib/utils');
+
+var VolumesList = require('./list');
+var VolumesForm = require('./volume-form');
+var FilterForm = require('./../../filter-form');
+
+var Volumes = require('../../../models/volumes');
+var Volume = require('../../../models/volume');
+
+var REQUEST_TIMEOUT = 120 * 1000;
```
Comment (dan-luk) [inline www/js/components/pages/volumes/index.jsx]: is it specific for volumes?

File: www/js/components/pages/volumes/index.jsx [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+    getInitialState: function () {
+        return {
+            filterTypes: ['name', 'owner_uuid', 'type', 'volume_state', 'tag']
+        };
+    },
+    statics: {
+        sidebar: 'volumes',
+        url: function () {
+            var url = 'volumes';
+            return location.pathname === '/volumes' ? (url + location.search || '') : url;
```
Comment (dan-luk) [inline www/js/components/pages/volumes/index.jsx]: CONST 'volumes'

why `|| ''`?

File: www/js/components/pages/volumes/list.jsx [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+    componentDidMount: function () {
+        var volumes = this.state.volumes;
+        volumes.on('reset', this.onSync, this);
+        volumes.on('add', this.onSync, this);
+        volumes.on('remove', this.onSync, this);
+        volumes.on('sync', this.onSync, this);
+        volumes.on('error', function (model, xhr) {
+            var appendedText = '';
+            if (xhr.statusText === 'timeout') {
+                appendedText = 'Connection timed out - please try again.';
```
Comment (dan-luk) [inline www/js/components/pages/volumes/list.jsx]: appendedText -> errorMessage

File: www/js/components/pages/volumes/list.jsx [hunk 16]
```diff
@@ -0,0 +1,146 @@
...
+            event.preventDefault();
+            event.stopPropagation();
+        }
+        var volume = model.toJSON();
+        var confirm = window.confirm(
+            _.str.sprintf('Are you sure you want to delete the volume "%s" ?', volume.name)
+        );
+        if (confirm) {
+            model.destroy({contentType: 'application/json', data: JSON.stringify(volume), wait: true}).done(function () {
+                var notifyMsg = _.str.sprintf('Volume <strong>%s</strong> deleted successfully.', volume.name);
```
Comment (dan-luk) [inline www/js/components/pages/volumes/list.jsx]: has been deleted

File: www/js/components/pages/volumes/list.jsx [hunk 17]
```diff
@@ -0,0 +1,146 @@
...
+        return (
+            <tr key={volume.uuid}>
+                <td className="status">
+                    <span className={'state ' + volume.state}>{volume.state}</span>
+                </td>
+                <td className="name">
+                    <a onClick={this.showVolume.bind(this, model)}>{volume.name || volume.uuid}</a>
+                </td>
+                <td className="type">
+                    {volume.type || null}
```
Comment (dan-luk) [inline www/js/components/pages/volumes/list.jsx]: null?

File: www/js/components/pages/volumes/networks.jsx [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+
+        if (this.state.error) {
+            return <div className="volume-networks">
+                <div className="loading error">
+                    <h1><i className="fa fa-warning" /> Networks could not be retrieved</h1>
+                    <p><code>{this.state.error}</code></p>
+                </div>
+            </div>;
+        }
+
```
Comment (dan-luk) [inline www/js/components/pages/volumes/networks.jsx]: if no networks?

File: www/js/components/pages/volumes/networks.jsx [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+                <div className="loading">
+                    <i className="fa fa-circle-o-notch fa-spin" /> Fetching Networks
+                </div>
+            </div>;
+        }
+
+        if (this.state.error) {
+            return <div className="volume-networks">
+                <div className="loading error">
+                    <h1><i className="fa fa-warning" /> Networks could not be retrieved</h1>
```
Comment (dan-luk) [inline www/js/components/pages/volumes/networks.jsx]: Failed to retrieve networks.

File: www/js/components/pages/volumes/volume-form.jsx [hunk 20]
```diff
@@ -0,0 +1,221 @@
...
+        this.userInput = new UserInput({
+            accountsOnly: true,
+            el: React.findDOMNode(self.refs.owner)
+        });
+
+        this.userInput.render();
+        this.userInput.on('selected', self.onSelectUser, self);
+    },
+    onSelectUser: function () {
+        var selectedUser = this.userInput.selectedUser && this.userInput.selectedUser.id;
```
Comment (dan-luk) [inline www/js/components/pages/volumes/volume-form.jsx]: selectedUser -> userId?

File: www/js/components/pages/volumes/volume-form.jsx [hunk 21]
```diff
@@ -0,0 +1,221 @@
...
+        selectedUser = selectedUser && selectedUser.length === 36 ? selectedUser : null;
+        var ownerUuid =this.state.owner_uuid;
```
Comment (dan-luk) [inline www/js/components/pages/volumes/volume-form.jsx]: space

File: www/js/components/pages/volumes/volume-form.jsx [hunk 22]
```diff
@@ -0,0 +1,221 @@
...
+        this.setState({error: null, creating: true});
+        api.post('/api/volumes').send(data).end(function (res) {
+            if (res.error) {
+                this.setState({error: res.body, creating: false});
+                return;
+            }
+            if (res.ok) {
+                app.vent.trigger('notification', {
+                    level: 'success',
+                    message: _.str.sprintf('Volume <strong>%s</strong> created successfully', state.name)
```
Comment (dan-luk) [inline www/js/components/pages/volumes/volume-form.jsx]: has been created

File: www/js/components/pages/volumes/volume.jsx [hunk 23]
```diff
@@ -0,0 +1,178 @@
...
+                data = data.get('uuid');
+            }
+
+            return _.str.sprintf('/volumes/%s', data);
+        }
+    },
+    getInitialState: function () {
+        var volume = this.props.volume;
+        return {
+            volume: volume && typeof volume === 'object' ? volume : new VolumeModel({uuid: this.props.uuid || volume}),
```
Comment (dan-luk) [inline www/js/components/pages/volumes/volume.jsx]: extract var for `volume && typeof volume === 'object'`

File: www/js/components/pages/volumes/volume.jsx [hunk 24]
```diff
@@ -0,0 +1,178 @@
...
+                <div className="loading">
+                    <i className="fa fa-circle-o-notch fa-spin" /> Fetching Volume
+                </div>
+            </div>;
+        }
+
+        if (this.state.error) {
+            return <div id="page-user">
+                <div className="loading error">
+                    <h1><i className="fa fa-warning" /> Volume could not be retrieved</h1>
```
Comment (dan-luk) [inline www/js/components/pages/volumes/volume.jsx]: "can not"

File: www/js/components/pages/volumes/volume.jsx [hunk 25]
```diff
@@ -0,0 +1,178 @@
...
+                </div>
+            </div>;
+        }
+
+        if (!this.state.volume) {
+            return <div id="page-volume">
+                <div className="page-header">
+                    <h2>Volume Not Found</h2>
+                </div>
+                <p>The Volume with ID <code>{this.props.uuid}</code> could not be found.</p>
```
Comment (dan-luk) [inline www/js/components/pages/volumes/volume.jsx]: can not

File: lib/adminui.js (hunk 9 above)
Comment (andyYatskov) [inline lib/adminui.js]: all parentheses need here.

Comment (dan-luk) [inline lib/adminui.js]: You are right!
But space is left?

Comment (dan-luk) [inline lib/adminui.js]: why this is not done

File: www/js/components/pages/volumes/volume-form.jsx [hunk 26]
```diff
@@ -117,60 +117,61 @@ var VolumeForm = React.createClass({
...
-                        <form className="form-horizontal">
-                            {this.state.error && <ErrorAlert error={state.error} />}
-                            <div className="form-group">
-                                <label className="control-label col-sm-4">Name</label>
-                                <div className="controls col-sm-7">
-                                    <input placeholder="name of volume" type="text" onChange={this._handleChangeName} value={state.name} className="form-control" name="name" />
-                                </div>
+            <div className="panel panel-info">
+                <div className="panel-heading">
+                    Create a new Volume
```
Comment (dan-luk) [inline www/js/components/pages/volumes/volume-form.jsx]: Create new Volume

**Pull request closed without merging**